### PR TITLE
Stop using FileSystem and Common submodules

### PR DESCRIPTION
**Pull request by jeffkl**
## Bug

- [x] Fixes: https://github.com/NuGet/Home/issues/2345

Regression? Last working version:

## Description
Most of this was done by @nkolev92, I just continued his work.  Removes the FileSystem and Common submodules in favor of package references.  This gets rid of the classes being compiled into our assemblies so they don't conflict with other types.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -1,11 +1,3 @@
-[submodule "submodules/FileSystem"]
```
Comment (nkolev92) [inline .gitmodules:1]: This is a 3 year old dream :D Glad it's finally done.

**Review by nkolev92 — DISMISSED** [review]
Love it. 

Did RPS pass?
Not sure if we need to do anything special to ensure source build is good beyond the build passing given that we're adding new assemblies. 

cc @crummel 

cc @zivkan We're deleting lots of public APIs here, but I think this is a risk we're 100% willing to take?

**Review by erdembayar — DISMISSED** [review]

**Review by crummel — DISMISSED** [review]
source-build already includes MS.Extensions so this should be fine by us.  Thanks for the ping!

File: src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude [hunk 2]
```diff
@@ -34,5 +34,8 @@ NuGet.Versioning.dll
 
 # 3rd party dlls to keep
 Lucene.Net.dll
+Microsoft.Extensions.FileProviders.Abstractions.dll
+Microsoft.Extensions.FileSystemGlobbing.dll
+Microsoft.Extensions.Primitives.dll
```
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude:39]: In nuget.exe's ilmerge config, there's also System.Memory, System.Buffers, and others.  These assemblies are already available in VS's PrivateAssemblies directory (and devenv.exe.config) when in VS. However, is there risk of any msbuild.exe command line scenario breaking?  Or does msbuild.exe also add VS's PrivateAssemblies path to the assembly load path?

The only file globbing codepaths I can think of are push and pack. Push doesn't exist in msbuild.exe. However, if a customer has a legacy csproj, and they use `-t:pack`, but they specify their own nuspec, which has file globbing, I think this is a codepath that is at risk.  Do we have any tests for such an edge case?

**Review by zivkan — DISMISSED** [review]
> We're deleting lots of public APIs here, but I think this is a risk we're 100% willing to take?

I agree. Removing the submodules will unavoidably break ABI compat, but since file globbing isn't a primary feature of nuget, we can just hope that very few people took that dependency.

Unfortunately namespaces changed, so it also breaks API compat. But that was the ASP.NET team's choice when they changed the namespace of these APIs.

**Comment by aortiz-msft** [comment]
So cool. Out of curiosity, does this reduce the build time, both locally and on our CI? If yes, by how much?

**Comment by jeffkl** [comment]
# RESOLVED

Does anyone know why only this one test is failing?  Or at least how to figure out why its failing?  Thanks in advance!

## NetCoreToolsVSandMSBuildNoOp
```
System.Management.Automation.ItemNotFoundException: Cannot find path 'C:\Users\TestUser\.nuget\packages\.tools\Microsoft.VisualStudio.Web.CodeGeneration.Tools\1.0.0-msbuild3-final\netcoreapp2.0\Microsoft.VisualStudio.Web.CodeGeneration.Tools.nuget.cache' because it does not exist.
   at System.Management.Automation.LocationGlobber.ExpandMshGlobPath(String path, Boolean allowNonexistingPaths, PSDriveInfo drive, ContainerCmdletProvider provider, CmdletProviderContext context)
   at System.Management.Automation.LocationGlobber.ResolveDriveQualifiedPath(String path, CmdletProviderContext context, Boolean allowNonexistingPaths, CmdletProvider& providerInstance)
   at System.Management.Automation.LocationGlobber.GetGlobbedMonadPathsFromMonadPath(String path, Boolean allowNonexistingPaths, CmdletProviderContext context, CmdletProvider& providerInstance)
   at System.Management.Automation.LocationGlobber.GetGlobbedProviderPathsFromMonadPath(String path, Boolean allowNonexistingPaths, CmdletProviderContext context, ProviderInfo& provider, CmdletProvider& providerInstance)
   at System.Management.Automation.SessionStateInternal.GetProperty(String[] paths, Collection`1 providerSpecificPickList, CmdletProviderContext context)
   at Microsoft.PowerShell.Commands.GetItemPropertyCommand.ProcessRecord()
```

https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=5964223&view=ms.vss-test-web.build-test-results-tab&runId=36869364&resultId=100100&paneView=debug

**Comment by nkolev92** [comment]
Did you try running the test locally?

There's some end to end instructions in the one note.

**Comment by jeffkl** [comment]
> Did you try running the test locally?
> 
> There's some end to end instructions in the one note.

Thanks for the pointer, I was able to run them locally after some trial and error, turns out we were missing a binding redirect entry in the pkgdef.  That fixed the problem locally, I've pushed the change.

**Review by nkolev92 — DISMISSED** [review]

File: src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude (hunk 2 above)
Comment (jeffkl) [reply src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude:39]: I'll do an experimental insertion, install it, and play around with some scenarios like this.

File: test/TestUtilities/Test.Utility/TestDotnetCLiUtility.cs [hunk 3]
```diff
@@ -233,6 +236,21 @@ private static void CopyRestoreArtifacts(string artifactsDirectory, string pathT
...
+
+            File.Copy(
+                sourceFileName: typeof(IFileInfo).Assembly.Location,
+                destFileName: Path.Combine(pathToSdkInCli, Path.GetFileName(typeof(IFileInfo).Assembly.Location)),
+                overwrite: true);
+
+            File.Copy(
+                sourceFileName: typeof(IChangeToken).Assembly.Location,
+                destFileName: Path.Combine(pathToSdkInCli, Path.GetFileName(typeof(IChangeToken).Assembly.Location)),
+                overwrite: true);
```
Comment (zivkan) [inline test/TestUtilities/Test.Utility/TestDotnetCLiUtility.cs]: Please put comments saying what files these are for. Or maybe remove the `using` statements at the top of the file, so we can see the namespace in the `typeof`.

FWIW, in the .NET 6 SDK, I already see the FilesystemGlobbing.dll, so if https://github.com/NuGet/Client.Engineering/issues/1271 gets completed, these file copies will probably not be needed.

File: src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude (hunk 2 above)
Comment (zivkan) [reply src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude:39]: An insertion is really nothing more than installing NuGet.Tools.vsix. You can test, without a fake insertion, just by downloading the vsix from the build artifacts, and installing that on the VM. You just need to run the vspreinstall script to use the fake signed, PR build.

**Review by zivkan — DISMISSED** [review]

File: src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude (hunk 2 above)
Comment (jeffkl) [reply src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude:39]: I must be doing something wrong.  I did the following:

1. Install latest internal preview
2. Verified NuGet Package Manager works
3. Run vspreinstall script
4. Download NuGet.Tools.vsix from my build
5. Install NuGet.Tools.vsix

Now I get an error in ActivityLog and NuGet package manager won't load

```
SetSite failed for package [NuGetPackage]Source: 'System.ComponentModel.Composition' Description: The composition produced multiple composition errors, with 11 root causes. The root causes are provided below. Review the CompositionException.Errors property for more detailed information.
1) No exports were found that match the constraint: ContractName NuGet.VisualStudio.IConsoleStatus RequiredTypeIdentity NuGet.VisualStudio.IConsoleStatus
    Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.ConsoleStatus (ContractName="NuGet.VisualStudio.IConsoleStatus")' on part 'NuGetVSExtension.NuGetPackage'.
    Element: NuGetVSExtension.NuGetPackage.ConsoleStatus (ContractName="NuGet.VisualStudio.IConsoleStatus") --> NuGetVSExtension.NuGetPackage
2) No exports were found that match the constraint: ContractName NuGet.PackageManagement.IDeleteOnRestartManager RequiredTypeIdentity NuGet.PackageManagement.IDeleteOnRestartManager
    Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.DeleteOnRestartManager (ContractName="NuGet.PackageManagement.IDeleteOnRestartManager")' on part 'NuGetVSExtension.NuGetPackage'.
    Element: NuGetVSExtension.NuGetPackage.DeleteOnRestartManager (ContractName="NuGet.PackageManagement.IDeleteOnRestartManager") --> NuGetVSExtension.NuGetPackage
3) No exports were found that match the constraint: ContractName NuGet.PackageManagement.VisualStudio.INuGetUILogger RequiredTypeIdentity NuGet.PackageManagement.VisualStudio.INuGetUILogger
    Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.OutputConsoleLogger (ContractName="NuGet.PackageManagement.VisualStudio.INuGetUILogger")' on part 'NuGetVSExtension.NuGetPackage'.
    Element: NuGetVSExtension.NuGetPackage.OutputConsoleLogger (ContractName="NuGet.PackageManagement.VisualStudio.INuGetUILogger") --> NuGetVSExtension.NuGetPackage
4) No exports were found that match the constraint: ContractName NuGet.ProjectManagement.INuGetProjectContext RequiredTypeIdentity NuGet.ProjectManagement.INuGetProjectContext
    Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.ProjectContext (ContractName="NuGet.ProjectManagement.INuGetProjectContext")' on part 'NuGetVSExtension.NuGetPackage'.
	Element: NuGetVSExtension.NuGetPackage.ProjectContext (ContractName="NuGet.ProjectManagement.INuGetProjectContext") --> NuGetVSExtension.NuGetPackage
5) No exports were found that match the constraint: ContractName NuGet.Configuration.ISettings RequiredTypeIdentity NuGet.Configuration.ISettings
    Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.Settings (ContractName="NuGet.Configuration.ISettings")' on part 'NuGetVSExtension.NuGetPackage'.
	Element: NuGetVSExtension.NuGetPackage.Settings (ContractName="NuGet.Configuration.ISettings") --> NuGetVSExtension.NuGetPackage
6) No exports were found that match the constraint: ContractName NuGet.PackageManagement.VisualStudio.IVsSolutionManager RequiredTypeIdentity NuGet.PackageManagement.VisualStudio.IVsSolutionManager
    Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.SolutionManager (ContractName="NuGet.PackageManagement.VisualStudio.IVsSolutionManager")' on part 'NuGetVSExtension.NuGetPackage'.
	Element: NuGetVSExtension.NuGetPackage.SolutionManager (ContractName="NuGet.PackageManagement.VisualStudio.IVsSolutionManager") --> NuGetVSExtension.NuGetPackage
7) No exports were found that match the constraint: ContractName NuGet.PackageManagement.UI.SolutionUserOptions RequiredTypeIdentity NuGet.PackageManagement.UI.SolutionUserOptions
    Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.SolutionUserOptions (ContractName="NuGet.PackageManagement.UI.SolutionUserOptions")' on part 'NuGetVSExtension.NuGetPackage'.
	Element: NuGetVSExtension.NuGetPackage.SolutionUserOptions (ContractName="NuGet.PackageManagement.UI.SolutionUserOptions") --> NuGetVSExtension.NuGetPackage
8) No exports were found that match the constraint: ContractName NuGet.VisualStudio.IVsSourceControlTracker RequiredTypeIdentity NuGet.VisualStudio.IVsSourceControlTracker
	Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.SourceControlTracker (ContractName="NuGet.VisualStudio.IVsSourceControlTracker")' on part 'NuGetVSExtension.NuGetPackage'.
	Element: NuGetVSExtension.NuGetPackage.SourceControlTracker (ContractName="NuGet.VisualStudio.IVsSourceControlTracker") --> NuGetVSExtension.NuGetPackage
9) No exports were found that match the constraint: ContractName NuGet.PackageManagement.UI.INuGetUIFactory RequiredTypeIdentity NuGet.PackageManagement.UI.INuGetUIFactory
	Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.UIFactory (ContractName="NuGet.PackageManagement.UI.INuGetUIFactory")' on part 'NuGetVSExtension.NuGetPackage'.
	Element: NuGetVSExtension.NuGetPackage.UIFactory (ContractName="NuGet.PackageManagement.UI.INuGetUIFactory") --> NuGetVSExtension.NuGetPackage
10) No exports were found that match the constraint: ContractName NuGet.VisualStudio.Common.IServiceBrokerProvider RequiredTypeIdentity NuGet.VisualStudio.Common.IServiceBrokerProvider
	Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.ServiceBrokerProvider (ContractName="NuGet.VisualStudio.Common.IServiceBrokerProvider")' on part 'NuGetVSExtension.NuGetPackage'.
	Element: NuGetVSExtension.NuGetPackage.ServiceBrokerProvider (ContractName="NuGet.VisualStudio.Common.IServiceBrokerProvider") --> NuGetVSExtension.NuGetPackage
11) No exports were found that match the constraint: ContractName NuGet.VisualStudio.INuGetExperimentationService RequiredTypeIdentity NuGet.VisualStudio.INuGetExperimentationService
	Resulting in: Cannot set import 'NuGetVSExtension.NuGetPackage.NuGetExperimentationService (ContractName="NuGet.VisualStudio.INuGetExperimentationService")' on part 'NuGetVSExtension.NuGetPackage'.
	Element: NuGetVSExtension.NuGetPackage.NuGetExperimentationService (ContractName="NuGet.VisualStudio.INuGetExperimentationService") --> NuGetVSExtension.NuGetPackage
```

End-to-end and Apex tests passed so the VSIX isn't completely hosed right?  I must be doing something stupid.

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude:39]: Weird. Maybe force rebuild the mef cache? 
https://marketplace.visualstudio.com/items?itemName=MadsKristensen.ClearMEFComponentCache#:~:text=Using%20the%20extension&text=Clicking%20the%20Clear%20MEF%20Component,cause%20any%20unwanted%20side%20effects.

Comment (jeffkl) [reply src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixinclude:39]: Couldn't get this to work so I just did an experimental insertion so I could install the build.

ClassLibrary1.csproj
```xml

<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net472</TargetFramework>
    <NuspecFile>ClassLibrary1.nuspec</NuspecFile>
  </PropertyGroup>
</Project>
```
ClassLibrary1.nuspec
```xml
<?xml version="1.0" encoding="utf-8" ?>
<package xmlns="http://schemas.microsoft.com/packaging/2010/07/nuspec.xsd">
  <metadata>
    <id>ClassLibrary1</id>
    <version>1.0.0</version>
    <description>ClassLibrary1</description>
    <authors>Me</authors>
    <dependencies>
      <group targetFramework="net472" />
    </dependencies>
  </metadata>
  <files>
    <file src="bin\Debug\**\*.dll" target="lib" />
  </files>
</package>
```

Tests:

Downloaded `nuget.exe` from my build drop:
```
D:\temp\ClassLibrary1>"C:\Users\jeffkl\Downloads\NuGet.exe" pack ClassLibrary1.nuspec
Attempting to build package from 'ClassLibrary1.nuspec'.
Successfully created package 'D:\temp\ClassLibrary1\ClassLibrary1.1.0.0.nupkg'.
```

Downloaded `NuGet.Build.Tasks.Pack.6.3.0-preview.1.11.nupkg` from my build drop and overlayed it on top of `C:\Program Files\dotnet\sdk\6.0.300-preview.22178.8\Sdks\NuGet.Build.Tasks.Pack\Desktop`:
```
D:\temp\ClassLibrary1>del *.nupkg /s &msbuild /t:Build;Pack /bl /restore /clp:v=m;summary;forcenoalign /m /nr:False
Deleted file - D:\temp\ClassLibrary1\ClassLibrary1.1.0.0.nupkg
Deleted file - D:\temp\ClassLibrary1\bin\Debug\ClassLibrary1.1.0.0.nupkg
Microsoft (R) Build Engine version 17.2.0-preview-22178-04+ce7523296 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

D:\VS2022-Exp\MSBuild\Current\Bin\amd64\msbuild.exe /bl /clp:v=m;summary;forcenoalign /m /nr:False /restore /t:Build;Pack .\ClassLibrary1.sln
  Determining projects to restore...
  All projects are up-to-date for restore.
  You are using a preview version of .NET. See: https://aka.ms/dotnet-core-preview
  ClassLibrary1 -> D:\temp\ClassLibrary1\bin\Debug\net472\ClassLibrary1.dll
  Successfully created package 'D:\temp\ClassLibrary1\bin\Debug\ClassLibrary1.1.0.0.nupkg'.

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:02.11
```
image

I also checked the binding redirects in `devenv.exe.config` and they do cover `System.Memory`, `System.Buffers`, etc.  

`MSBuild.exe.config` also has the binding redirects which match Visual Studio's:
```xml
      <dependentAssembly>
        <assemblyIdentity name="System.Buffers" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.3.0" newVersion="4.0.3.0" />
        <codeBase version="4.0.3.0" href="..\System.Buffers.dll" />
      </dependentAssembly>
      <!-- Pull plugins that reference SCI up to our version in case they depended on our copy of the older version -->
      <dependentAssembly>
        <assemblyIdentity name="System.Collections.Immutable" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-5.0.0.0" newVersion="5.0.0.0" />
        <codeBase version="5.0.0.0" href="..\System.Collections.Immutable.dll" />
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="System.Memory" publicKeyToken="cc7b13ffcd2ddd51" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.1.1" newVersion="4.0.1.1" />
        <codeBase version="4.0.1.1" href="..\System.Memory.dll" />
      </dependentAssembly>
      <dependentAssembly>
        <assemblyIdentity name="System.Runtime.CompilerServices.Unsafe" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-6.0.0.0" newVersion="6.0.0.0" />
        <codeBase version="6.0.0.0" href="..\System.Runtime.CompilerServices.Unsafe.dll" />
      </dependentAssembly>
```

`NuGet.exe` does end up with a reference to `System.Runtime.CompilerServices.Unsafe` but its not ILMerge'd into it so any code path that ends up needing it would fail. 

image

So everything appears to be working but you could consider it to be brittle.  Should I add more assemblies to the ILMerge?  This list was just what @nkolev92 had done in his change so I used it in mine...

**Comment by jeffkl** [comment]
> So cool. Out of curiosity, does this reduce the build time, both locally and on our CI? If yes, by how much?

@aortiz-msft It will reduce the build time by a little since its not fetching these submodules during build anymore and no longer compiling the sources into our assemblies.  I doubt it's going to save minutes of time though.  This change is more about reducing the overhead of cloning and building our repository.  There's also a bug where the types from these classes are `public` and can mess up consumers of our assemblies.

File: test/TestUtilities/Test.Utility/TestDotnetCLiUtility.cs (hunk 3 above)
Comment (jeffkl) [reply test/TestUtilities/Test.Utility/TestDotnetCLiUtility.cs]: Done

**Comment by jeffkl** [comment]
@nkolev92 or @zivkan do you know if I need to update this?

https://github.com/NuGet/NuGet.Client/blob/5b0ee2562df6cf2c66cc9905a8a65f2b09ea5473/setup/Microsoft.VisualStudio.NuGet.BuildTools.swr#L7-L29

**Comment by nkolev92** [comment]
@jeffkl Yep. 

That VSIX is for the build tools SKU and it needs restore to work.

**Comment by jeffkl** [comment]
> That VSIX is for the build tools SKU and it needs restore to work.

@nkolev92 do you know if Apex, end-to-end tests, or anything else do anything with BuildTools SKU?  Or do I just need to manually test it?

**Comment by nkolev92** [comment]
They don't, but there are manual vendors tests for this, but they are not super detailed.

**Comment by jeffkl** [comment]
@NuGet/nuget-client please review

**Review by nkolev92 — APPROVED** [review]
🚢

**Comment by nkolev92** [comment]
🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳
